File: prototype/lib/moof/pattern_matcher.rb
module Moof
  module PatternMatcher
    # Returns {success: true/false, bindings: {name => value}}
    def self.match(pattern, value, interpreter)
      case pattern
      when AST::MatchWildcard
        {success: true, bindings: {}}
      when AST::MatchBind
        {success: true, bindings: {pattern.name => value}}
      when AST::IntegerLiteral, AST::FloatLiteral, AST::StringLiteral, AST::BoolLiteral
        {success: value == pattern.value, bindings: {}}
      when AST::NilLiteral
        {success: value.nil?, bindings: {}}
      when AST::MatchList
        match_list(pattern, value, interpreter)
      when AST::MatchMap
        match_map(pattern, value, interpreter)
      when AST::MatchConstructor
        match_constructor(pattern, value, interpreter)
      when AST::Identifier
        if pattern.name == "_"
          {success: true, bindings: {}}
        else
          {success: true, bindings: {pattern.name => value}}
        end
      else
        {success: false, bindings: {}}
      end
    end

    private

    def self.match_list(pattern, value, _interpreter)
      return {success: false, bindings: {}} unless value.is_a?(Array)

      elements = pattern.elements
      rest = pattern.rest

      if rest
        # With rest pattern: need at least as many elements as non-rest patterns
        return {success: false, bindings: {}} if value.length < elements.length
      else
        # Without rest: exact length match
        return {success: false, bindings: {}} if value.length != elements.length
      end

      bindings = {}
      elements.each_with_index do |el_pattern, i|
        result = match(el_pattern, value[i], _interpreter)
        return {success: false, bindings: {}} unless result[:success]
        bindings.merge!(result[:bindings])
      end

      if rest
        bindings[rest] = value[elements.length..] || []
      end

      {success: true, bindings: bindings}
    end

    def self.match_map(pattern, value, _interpreter)
      return {success: false, bindings: {}} unless value.is_a?(Hash)

      bindings = {}
      pattern.pairs.each do |key, sub_pattern|
        return {success: false, bindings: {}} unless value.key?(key)
        result = match(sub_pattern, value[key], _interpreter)
        return {success: false, bindings: {}} unless result[:success]
        bindings.merge!(result[:bindings])
      end

      {success: true, bindings: bindings}
    end

    def self.match_constructor(pattern, value, _interpreter)
      return {success: false, bindings: {}} unless value.is_a?(MoofObject)
      return {success: false, bindings: {}} unless value.klass.name == pattern.class_name

      bindings = {}
      field_names = value.klass.fields
      pattern.bindings.each_with_index do |bind_pattern, i|
        field_name = field_names[i]
        return {success: false, bindings: {}} unless field_name
        field_value = value.fields[field_name]
        result = match(bind_pattern, field_value, _interpreter)
        return {success: false, bindings: {}} unless result[:success]
        bindings.merge!(result[:bindings])
      end

      {success: true, bindings: bindings}
    end
  end
end

File: src/pattern_matcher.rs
use crate::ast::{Expr, Pattern};
use crate::interpreter::Interpreter;
use crate::value::Value;

pub struct MatchResult {
    pub success: bool,
    pub bindings: Vec<(String, Value)>,
}

impl MatchResult {
    fn success(bindings: Vec<(String, Value)>) -> Self {
        MatchResult { success: true, bindings }
    }

    fn failure() -> Self {
        MatchResult { success: false, bindings: vec![] }
    }
}

pub fn match_pattern(pattern: &Pattern, value: &Value, interp: &Interpreter) -> MatchResult {
    match pattern {
        Pattern::Wildcard => MatchResult::success(vec![]),

        Pattern::Bind(name) => {
            if name == "_" {
                MatchResult::success(vec![])
            } else {
                MatchResult::success(vec![(name.clone(), value.clone())])
            }
        }

        Pattern::Literal(expr) => {
            let matches = match expr.as_ref() {
                Expr::Integer(n, _) => matches!(value, Value::Integer(v) if *v == *n),
                Expr::Float(n, _) => matches!(value, Value::Float(v) if *v == *n),
                Expr::Str(s, _) => matches!(value, Value::Str(v) if v == s),
                Expr::Bool(b, _) => matches!(value, Value::Bool(v) if *v == *b),
                Expr::Nil(_) => matches!(value, Value::Nil),
                _ => false,
            };
            if matches {
                MatchResult::success(vec![])
            } else {
                MatchResult::failure()
            }
        }

        Pattern::List(elements, rest) => match_list(elements, rest.as_deref(), value, interp),
        Pattern::Map(pairs) => match_map(pairs, value, interp),
        Pattern::Constructor(class_name, bindings) => {
            match_constructor(class_name, bindings, value, interp)
        }
    }
}

fn match_list(
    elements: &[Pattern],
    rest: Option<&Pattern>,
    value: &Value,
    interp: &Interpreter,
) -> MatchResult {
    let items = match value {
        Value::List(items) => items,
        _ => return MatchResult::failure(),
    };

    if rest.is_some() {
        if items.len() < elements.len() {
            return MatchResult::failure();
        }
    } else if items.len() != elements.len() {
        return MatchResult::failure();
    }

    let mut bindings = Vec::new();

    for (i, pat) in elements.iter().enumerate() {
        let result = match_pattern(pat, &items[i], interp);
        if !result.success {
            return MatchResult::failure();
        }
        bindings.extend(result.bindings);
    }

    if let Some(rest_pat) = rest {
        let rest_value = Value::List(items[elements.len()..].to_vec());
        let result = match_pattern(rest_pat, &rest_value, interp);
        if !result.success {
            return MatchResult::failure();
        }
        bindings.extend(result.bindings);
    }

    MatchResult::success(bindings)
}

fn match_map(
    pairs: &[(String, Pattern)],
    value: &Value,
    interp: &Interpreter,
) -> MatchResult {
    let map_pairs = match value {
        Value::Map(pairs) => pairs,
        _ => return MatchResult::failure(),
    };

    let mut bindings = Vec::new();

    for (key, sub_pattern) in pairs {
        let found = map_pairs.iter().find(|(k, _)| k == key);
        match found {
            Some((_, val)) => {
                let result = match_pattern(sub_pattern, val, interp);
                if !result.success {
                    return MatchResult::failure();
                }
                bindings.extend(result.bindings);
            }
            None => return MatchResult::failure(),
        }
    }

    MatchResult::success(bindings)
}

fn match_constructor(
    class_name: &str,
    pat_bindings: &[Pattern],
    value: &Value,
    interp: &Interpreter,
) -> MatchResult {
    let obj = match value {
        Value::Object(obj) => obj,
        _ => return MatchResult::failure(),
    };

    let obj_class_name = obj.class.borrow().name.clone();
    if obj_class_name != class_name {
        return MatchResult::failure();
    }

    let field_names = obj.class.borrow().all_fields();
    let mut bindings = Vec::new();

    for (i, bind_pattern) in pat_bindings.iter().enumerate() {
        let field_name = match field_names.get(i) {
            Some(name) => name,
            None => return MatchResult::failure(),
        };
        let field_value = obj.fields.get(field_name).cloned().unwrap_or(Value::Nil);
        let result = match_pattern(bind_pattern, &field_value, interp);
        if !result.success {
            return MatchResult::failure();
        }
        bindings.extend(result.bindings);
    }

    MatchResult::success(bindings)
}
